rfcs/0000-error-handling.md

```diff
@@ -0,0 +1,154 @@
+
+## Urgency
+
+Sooner is better than later.
+
+## Terminology
+
+A new subgraph health status called "unhealthy" is added, so instead of a subgraph being failed or not, it can be healthy, unhealthy or failed.
+
+## Detailed Design
+
+### Error handling
+
+Subgraphs now have three possible health status: 'Healthy' is normal operation, 'Unhealthy' is syncing with errors, and 'Failed' is halted due to an error.
+
+If a WASM handler terminates with a trap, the subgraph becomes unhealthy. The node discards all entity operations and data source operations done by the handler and continues syncing.
+
+For subgraphs that are not yet synced, any handler error is still a failure and halts syncing, so errors in pending versions will not make their way to the current version. 
+
+When the node starts, we currently reset the failed flag on subgraphs. Failed subgraphs will still be reset to healthy to attempt to recover, but unhealthy subgraphs are kept unhealthy, since the subgraph will already be past the errors.
+
+Any graphql queries to an unhealthy or failed subgraph will include an error of the form:
+
+```
+{
+	message: "Subgraph not healthy"
+}
+```
+
+So that applications will always be aware that queries may have inconsistent data. Developers should refer to the indexing status API for details on the errors.
+
+### Indexing Status API
+
+The indexing status API will be revamped.
+
+The following top level query is added:
+
+    indexingStatusForSubgraphName(subgraphName: String!): SubgraphIndexingStatus
+
+Which returns null if the name is not found, and returns the status for the current version of the subgraph otherwise.
+
+A field `lastHealthyBlock`  is added to record the block previous to the first error.
+
+Errors are made more than just a string with the `Error` type.
+
+The complete indexing status API schema is:
+
+```graphql
+type Query {
+  indexingStatusForSubgraphName(subgraphName: String!): SubgraphIndexingStatus
+  indexingStatusesForSubgraphName(subgraphName: String!): [SubgraphIndexingStatus!]!
+  indexingStatuses(subgraphs: [String!]): [SubgraphIndexingStatus!]!
+}
+
+type SubgraphIndexingStatus {
+  subgraph: String!
+  synced: Boolean!
+  health: Health!
+  errors: [Error!]! # Sorted from first to last
+  chains: [ChainIndexingStatus!]!
+  node: String!
+}
+
+# Fields are moved from `EthereumIndexingStatus` to the interface.
+interface ChainIndexingStatus {
+  network: String!
+  chainHeadBlock: Block
+  earliestBlock: Block
+  latestBlock: Block
+  lastHealthyBlock: Block
+}
+
+type EthereumIndexingStatus implements ChainIndexingStatus { }
+
+# Renamed from `EthereumBlock` to `Block`
+type Block {
+  hash: Bytes!
+  number: BigInt!
+}
+
+type Error {
+	message: String!
+	
+	# Context for the error.
+	network: String!
+	block: Block!
+	handler: String
+}
+
+enum Health {
+  "Subgraph syncing normally"
+  Healthy,
+  "Subgraph syncing but with errors"
+  Unhealthy,
+  "Subgraph halted due to errors"
+  Failed,
```

Comment:
Nit: Let's use lower-case enum values. It feels more consistent with all other enums we use in our GraphQL APIs (e.g. `orderDirection: asc`, `where: { something_gt: ... }`).

Reply:
Done.